### PR TITLE
Adding a recipe for better knowledge distillation

### PR DESCRIPTION
**Pull request by sayakpaul**
Knowledge distillation is a very important technique to get right when it comes to using them for production systems. Because of its simplicity, it's often a preferred way to optimize larger models that may not be suitable for different resource-constrained production environments. Even if we have access to unlimited compute resources to host our models on the cloud, it's still a good engineering practice to aim for optimizing storage and other deployment costs without compromising performance. 

This recipe discussed the [Knowledge distillation: A good teacher is patient and consistent](https://arxiv.org/pdf/2106.05237.pdf) paper from Google Brain and demonstrates results on the Flowers102 dataset.

### CODE REVIEW

File: examples/keras_recipes/distillation_recipe.py [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+class Distiller(tf.keras.Model):
+    def __init__(self, student, teacher):
+        super(Distiller, self).__init__()
+        self.student = student
+        self.teacher = teacher
+
+    def compile(
+        self, optimizer, metrics, distillation_loss_fn, temperature=TEMPERATURE,
+    ):
+        super(Distiller, self).compile(optimizer=optimizer, metrics=metrics)
```
Comment (fchollet) [inline examples/keras_recipes/distillation_recipe.py:301]: You allow arbitrary metrics but don't take them into account in the training step (though you do in the test step). You should add a `self.compiled_metrics.update_state(y, student_predictions)` in `train_step`.

File: examples/keras_recipes/distillation_recipe.py [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+        # Compute gradients
+        trainable_vars = self.student.trainable_variables
+        gradients = tape.gradient(distillation_loss, trainable_vars)
+
+        # Update weights
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+
+        # Report progress
+        results = {m.name: m.result() for m in self.metrics}
+        results.update({"distillation_loss": distillation_loss})
```
Comment (fchollet) [inline examples/keras_recipes/distillation_recipe.py:331]: You should add a `Mean` metric to track this quantity, listed in `self.metrics`, otherwise you're only displaying the value from the last batch, not the running average. You can have `self.metrics` return the super value, plus your custom metric (basically mixing the framework-managed metrics and your own).

This makes me think that it could be useful to add automated tracking of metrics on `Model`, as a future Keras feature

File: examples/keras_recipes/distillation_recipe.py [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+        return results
```
Comment (fchollet) [inline examples/keras_recipes/distillation_recipe.py:332]: If everything is listed in `metrics` you can just return `{m.name: m.result() for m in self.metrics}`

File: examples/keras_recipes/distillation_recipe.py [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+
+A warmup cosine learning rate schedule is used in the paper. This schedule is also
+typical for many pre-training methods especially for computer vision.
+"""
+
+# Some code is referred from here:
+# https://www.kaggle.com/ashusma/training-rfcx-tensorflow-tpu-effnet-b2.
+
+
+class WarmUpCosine(keras.optimizers.schedules.LearningRateSchedule):
```
Comment (fchollet) [inline examples/keras_recipes/distillation_recipe.py:366]: Nice, it's great to show the use of a custom LR schedule!

File: examples/keras_recipes/distillation_recipe.py [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+        self.learning_rate_base = learning_rate_base
+        self.total_steps = total_steps
+        self.warmup_learning_rate = warmup_learning_rate
+        self.warmup_steps = warmup_steps
+        self.pi = tf.constant(np.pi)
+
+    def __call__(self, step):
+        if self.total_steps < self.warmup_steps:
+            raise ValueError("Total_steps must be larger or equal to warmup_steps.")
+        learning_rate = (
```
Comment (fchollet) [inline examples/keras_recipes/distillation_recipe.py:381]: That's a big paragraph-line to define the LR, consider breaking it up into a few separate lines / intermediate variables to make it more readable.

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR! It's looking great!

File: examples/keras_recipes/distillation_recipe.py [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+BiT variants of ResNets use Group Normalization ([Wu et al.](https://arxiv.org/abs/1803.08494))
+and Weight Standardization ([Qiao et al.](https://arxiv.org/abs/1903.10520v2))
+in place of Batch Normalization ([Ioffe et al.](https://arxiv.org/abs/1502.03167)).
+In order to keep the runtime of this example short, we will be using a BiT
+ResNet101x3 trained on the Flowers102 dataset. However, you can refer to
+[this notebook](https://github.com/sayakpaul/FunMatch-Distillation/blob/main/train_bit.ipynb)
+to know more about the training process. This model produces 98.18% top-1 accuracy on the
+test set of Flowers102.
+
+The model weights are hosted on Kaggle as a dataset. To download the weights you can do
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/keras_recipes/distillation_recipe.py:111]: I had thought about that but the teacher model is 2.9 GB and GitHub won't allow it. The limit is 2 GB.

File: examples/keras_recipes/distillation_recipe.py (hunk 1 above)
Comment (sayakpaul) [reply examples/keras_recipes/distillation_recipe.py:301]: Since the labels are not used for calculating any training metrics like accuracy, I didn't do it.

File: examples/keras_recipes/distillation_recipe.py [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+class Distiller(tf.keras.Model):
+    def __init__(self, student, teacher):
+        super(Distiller, self).__init__()
+        self.student = student
+        self.teacher = teacher
+        self.loss_tracker = keras.metrics.Mean(name="distillation_loss")
+
+    @property
+    def metrics(self):
+        return [self.loss_tracker]
```
Comment (fchollet) [inline examples/keras_recipes/distillation_recipe.py:301]: All of your metrics should be listed here. That includes the metrics created via `compile()`. So here you have to retrieve `super().metrics` and add to it your loss tracker.

File: examples/keras_recipes/distillation_recipe.py (hunk 1 above)
Comment (fchollet) [reply examples/keras_recipes/distillation_recipe.py:306]: Then the code is inconsistent. In your train_step you are returning results for all metrics, not just for the loss. Yet, you are not updating these metrics.

Either you just care about the loss, in which case there should not be other metrics, or you actually support arbitrary metrics, which requires updating them in `train_step` and listing them in `self.metrics`

File: examples/keras_recipes/distillation_recipe.py [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+        trainable_vars = self.student.trainable_variables
+        gradients = tape.gradient(distillation_loss, trainable_vars)
+
+        # Update weights
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+
+        # Report progress
+        self.loss_tracker.update_state(distillation_loss)
+        results = {m.name: m.result() for m in self.metrics}
+        results.update({"distillation_loss": self.loss_tracker.result()})
```
Comment (fchollet) [inline examples/keras_recipes/distillation_recipe.py:337]: The way the code is currently is currently written, these 2 lines are redundant. `self.metrics` just contains the loss tracker, so `results` contains `{"distillation_loss": self.loss_tracker.result()}`, and you're updating it at the next line with... the loss tracker result.

The problem is solved if you just list all metrics in `self.metrics` (and update the compiled metrics in `train_step`).

File: examples/keras_recipes/distillation_recipe.py [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+
+# Comes from Table 4 and "Training setup" section.
+TEMPERATURE = 10  # Will be used to soften the logits before they go to softmax.
+INIT_LR = 0.003  # Initial learning rate that will be decayed over the training period.
+WEIGHT_DECAY = 0.001  # Will be used for regularization.
+CLIP_THRESHOLD = 1.0  # Will be used for clipping the gradients w.r.t their L2-norm.
+
+BATCH_SIZE = 64
+BIGGER = 160
+RESIZE = 128
```
Comment (fchollet) [inline examples/keras_recipes/distillation_recipe.py:77]: BIGGER and RESIZE should be explained.

File: examples/keras_recipes/distillation_recipe.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/keras_recipes/distillation_recipe.py:111]: @fchollet is it possible to create a central GCS Bucket for community contributions by the Keras team?

File: examples/keras_recipes/distillation_recipe.py [hunk 10]
```diff
@@ -0,0 +1,519 @@
...
+class Distiller(tf.keras.Model):
+    def __init__(self, student, teacher):
+        super(Distiller, self).__init__()
+        self.student = student
+        self.teacher = teacher
+        self.loss_tracker = keras.metrics.Mean(name="distillation_loss")
+
+    @property
+    def metrics(self):
+        return super().metrics.append(self.loss_tracker)
```
Comment (fchollet) [inline examples/keras_recipes/distillation_recipe.py]: This is going to be returning `None`, because it returns the output of `list.append()`, which is None.

Do

```python
metrics = copy(super().metrics)  # Avoid modifying the original list
metrics.append(self.loss_tracker)
return metrics
```

**Review by fchollet — COMMENTED** [review]
Thanks for the update! I pushed the copyedits. One last issue, almost ready!

**Comment by sayakpaul** [comment]
@fchollet pushed all the generated files along with addressing the last issue. 

Let me know anything more is needed on my end.

**Review by fchollet — APPROVED** [review]
Awesome! Thanks for the great contribution 👍
